File: server/src/models/Incident.ts
import { Schema, model, models } from 'mongoose';

export const IncidentSchema = new Schema({
    incidentType: { type: String, required: true },
    description: { type: String, required: true },
    phoneNumber: { type: String },
    isUserVerified: { type: Boolean, default: false },
    showPhoneNo: { type: Boolean, default: false },
    isDeleted: { type: Boolean, default: false },
    location: {
        type: {
          type: String,
          enum: ['Point'], // 'location.type' must be 'Point'
          required: true,
          default: 'Point',
        },
        coordinates: {
          type: [Number],
          required: true,
        },
      },
});

IncidentSchema.index({ incidentType: 1, 'location.coordinates': 1 }, { unique: true });
// Create a geospatial index on the 'location' field
IncidentSchema.index({ location: '2dsphere' });

export const Incident = models.Incident ?? model('Incident', IncidentSchema);

File: app/lib/models/incident.dart
class Incident {
  final String id;
  final String incidentType;
  final String description;
  final String? phoneNumber;
  final bool isUserVerified;
  final bool showPhoneNo;
  final bool isDeleted;
  final double latitude;
  final double longitude;

  Incident({
    required this.id,
    required this.incidentType,
    required this.description,
    this.phoneNumber,
    required this.isUserVerified,
    required this.showPhoneNo,
    required this.isDeleted,
    required this.latitude,
    required this.longitude,
  });

  factory Incident.fromJson(Map<String, dynamic> json) {
    return Incident(
      id: json['_id']['\$oid'],
      incidentType: json['incidentType'],
      description: json['description'],
      phoneNumber: json['phoneNumber'],
      isUserVerified: json['isUserVerified'],
      showPhoneNo: json['showPhoneNo'],
      isDeleted: json['isDeleted'],
      latitude: json['location']['coordinates'][1],
      longitude: json['location']['coordinates'][0],
    );
  }
}
